File: TypeScript/src/session2.td/heritageMultiple/NatDecimal.ts
import {Nat} from "../../session1.td/Nat";
import {FabriqueNaturels} from "../../session1.td/FabriqueNaturels";
import {NombreDecimal} from "./NombreDecimal";
import {AlgebreNatDecimal, AlgebreNatDecimal_default} from "./AlgebreNatDecimal";
import {heriterInterfacesConcretes} from "./heritageInterfacesConcretes";

export class NatDecimal extends NombreDecimal implements AlgebreNatDecimal {

    public static FAB : FabriqueNaturels<Nat> = new NatDecimal("0");

    public constructor(rep : string) {
        super(rep);
    }


    private static nettoyer(s : string) : void{
        let debut : number = 0;
        let fin : number = 0;
        while((fin < s.length) && +(s.charAt(fin)) == 0){
            fin++;
        }
        s.replace(s.substring(debut, fin), "");
    }

    public creerNatAvecRepresentation(repDecimalecimale : string) : Nat {
        let repMutable : string = repDecimalecimale;
        NatDecimal.nettoyer(repMutable);
        repDecimalecimale = repMutable;
        if(repDecimalecimale == ""){
            repDecimalecimale = "0";
        }
        let min : string = "0";
        let max : string = "9";
        for(let i : number = 0; i < repDecimalecimale.length; i++){
            let c : string = repDecimalecimale.charAt(i);
            if(c < min)	throw new Error();
            if(c > max) throw new Error();
        }
        return new NatDecimal(repDecimalecimale);
    }

    public equals(x : Nat) : boolean {
        return this.val() == x.val();
    }

    public toString() : string {
        return this._chiffres;
    }

    div(x: Nat): Nat {
        return undefined;
    }

    modulo(x: Nat): Nat {
        return undefined;
    }

    produit(x: Nat): Nat {
        return undefined;
    }

    somme(x: Nat): Nat {
        return undefined;
    }

    un(): Nat {
        return undefined;
    }

    zero(): Nat {
        return undefined;
    }

}

heriterInterfacesConcretes(NatDecimal, [NombreDecimal, AlgebreNatDecimal_default]);

File: Java/src/session2/td/heritageMultiple/NatDecimal.java
package session2.td.heritageMultiple;

import session1.td.FabriqueNaturels;
import session1.td.Nat;

public class NatDecimal extends NombreDecimal implements AlgebreNatDecimal {


    public static final FabriqueNaturels<Nat> FAB = new NatDecimal("0");

    public NatDecimal(String rep) {
        super(rep);
    }


    private static void nettoyer(StringBuilder s){
        int debut = 0;
        int fin = 0;
        while((fin < s.length()) && Character.getNumericValue(s.charAt(fin)) == 0){
            fin++;
        }
        s.delete(debut, fin);
    }

    @Override
    public Nat creerNatAvecRepresentation(String repDecimalecimale) {
        StringBuilder repMutable = new StringBuilder(repDecimalecimale);
        nettoyer(repMutable);
        repDecimalecimale = new String(repMutable);
        if(repDecimalecimale.equals("")){
            repDecimalecimale = "0";
        }
        char min = Character.forDigit(0, 10);
        char max = Character.forDigit(9, 10);
        for(int i = 0; i < repDecimalecimale.length(); i++){
            char c = repDecimalecimale.charAt(i);
            if(c < min)	throw new IllegalArgumentException();
            if(c > max) throw new IllegalArgumentException();
        }
        return new NatDecimal(repDecimalecimale);
    }

    @Override
    public boolean equals(Object x){
        if(!(x instanceof Nat)) return false;
        Nat n = (Nat)x;
        return this.toString().equals(n.toString());
    }

    @Override
    public String toString() {
        return this.chiffres;
    }


}
